### PR TITLE
macros: expand `#[derive]`s after other attribute macros and improve intra-`#[derive]` ordering

### PR DESCRIPTION
**Pull request by jseyfried**
Fixes https://github.com/serde-rs/serde/issues/577.
cc #35900 
r? @alexcrichton

### CODE REVIEW

**Comment by alexcrichton** [comment]
@bors: r+

Thanks @jseyfried!

File: src/libsyntax_ext/deriving/mod.rs [hunk 1]
```diff
@@ -104,13 +105,40 @@ pub fn expand_derive(cx: &mut ExtCtxt,
         }
     };
 
-    if mitem.value_str().is_some() {
-        cx.span_err(mitem.span, "unexpected value in `derive`");
+    let mut derive_attrs = Vec::new();
+    item = item.map_attrs(|attrs| attrs.into_iter().filter_map(|attr| {
```
Comment (sgrif) [inline src/libsyntax_ext/deriving/mod.rs]: Could this use [`Iterator#partition`](https://doc.rust-lang.org/nightly/std/iter/trait.Iterator.html#method.partition) instead?

**Comment by jseyfried** [comment]
cc @nrc

**Pull request merged**